### PR TITLE
Feat: Add the `recursive_character` algorithm for Text chunking processor

### LINKED ISSUE #1261: [PROPOSAL] RecursiveCharacterTextSplitter as a new algorithm of Text Chunking Processor
### What are you proposing?
`RecursiveCharacterTextSplitter`
- ref: 
  - https://python.langchain.com/docs/how_to/recursive_text_splitter/
  - https://github.com/langchain-ai/langchain/blob/master/libs/text-splitters/langchain_text_splitters/character.py#L58

### What problems are you trying to solve?
I want to add a chunk processor to implement a recursive character text splitter, which is a popular chunking strategy. Text Chunking Processor in OpenSearch supports [fixed_token_length](https://opensearch.org/docs/latest/ingest-pipelines/processors/text-chunking/#fixed-token-length-algorithm) and [delimiter](https://opensearch.org/docs/latest/ingest-pipelines/processors/text-chunking/#delimiter-algorithm) for now, but the combination of these two cannot implement recursive character text splitter algorithm.

#### Are there any security considerations? 
No

#### Are there any breaking changes to the API
No

### What is the user experience going to be?
RecursiveCharacterTextSplitter will most likely help the developer who wants to split long texts in the format of  [Markdown](https://github.com/langchain-ai/langchain/blob/master/libs/text-splitters/langchain_text_splitters/character.py#L459), [Programming Language](https://github.com/langchain-ai/langchain/blob/master/libs/text-splitters/langchain_text_splitters/character.py#L196), or even [HTML](https://github.com/langchain-ai/langchain/blob/master/libs/text-splitters/langchain_text_splitters/character.py#L503).

#### Are there breaking changes to the User Experience?
No

### PR DESCRIPTION
**Pull request by YeonghyeonKO**
### Description
Add the recursive character chunker algorithm to TextChunking processors. It is designed for processing generic text, facilitating the division of larget text bodies into smaller, semantically meaningful chunks. With `delimiters` and `chunk_size` parameters, the algorithm attempts to split the text using these characters in sequential order until the resulting chunks satisfy the given size requirements.

The default list of delimiters used by the algorithm is:
- "\n\n": preserves paragraphs.
- "\n": preserves sentences when paragraphs are not feasible to keep intact.
- " ": keeps words together when sentences must be split.
- "": simply splits by character, ensuring the text is split even if no other delimiters are appropriate.

```
PUT _ingest/pipeline/text-chunking-ingest-pipeline
{
  "description": "A text chunking ingest pipeline using recursive character chunker",
  "processors": [
    {
      "text_chunking": {
        "algorithm": {
          "recursive_character": {
            "chunk_size": 50,
            "delimiters": ["\n\n", "\n", " ", ""]
          }
        },
        "field_map": {
          "passage_text": "passage_chunk"
        }
      }
    }
  ]
}
```

### Related Issues
Resolves #1261 ([PROPOSAL] RecursiveCharacterTextSplitter as a new algorithm of Text Chunking Processor)

### Check List
- [x] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md).
- [x] Commits are signed per the DCO using `--signoff`.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/neural-search/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/main/java/org/opensearch/neuralsearch/processor/chunker/RecursiveCharacterChunker.java [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+        this.delimiters = parseListOfStringWithDefault(parameters, DELIMITERS_FIELD, DEFAULT_DELIMITERS);
+        this.chunkSize = parsePositiveIntegerWithDefault(parameters, CHUNK_SIZE_FIELD, DEFAULT_CHUNK_SIZE);
+    }
+
+    @Override
+    public List<String> chunk(String content, Map<String, Object> runtimeParameters) {
+        int runtimeMaxChunkLimit = parseInteger(runtimeParameters, MAX_CHUNK_LIMIT_FIELD);
+        List<String> processedChunks = splitRecursive(content, delimiters, chunkSize, 0);
+
+        if (runtimeMaxChunkLimit > 0 && processedChunks.size() > runtimeMaxChunkLimit) {
```
Comment (yuye-aws) [inline src/main/java/org/opensearch/neuralsearch/processor/chunker/RecursiveCharacterChunker.java:43]: Still get an edge case where runtimeMaxChunkLimit  = 0. Can you use: **checkRunTimeMaxChunkLimit**?

File: src/main/java/org/opensearch/neuralsearch/processor/chunker/RecursiveCharacterChunker.java [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+            return limitedChunks;
+        } else {
+            return processedChunks;
+        }
+    }
+
+    private List<String> splitRecursive(String text, List<String> delimiters, int chunkSize, int level) {
+        if (text == null || text.isEmpty()) {
+            return new ArrayList<>();
+        }
```
Comment (yuye-aws) [inline src/main/java/org/opensearch/neuralsearch/processor/chunker/RecursiveCharacterChunker.java:75]: Since the content for the chunk method is always non-null, why this method needs to check null?

File: src/main/java/org/opensearch/neuralsearch/processor/chunker/RecursiveCharacterChunker.java [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+                }
+            }
+            limitedChunks.add(mergedTailBuilder.toString());
+            return limitedChunks;
+        } else {
+            return processedChunks;
+        }
+    }
+
+    private List<String> splitRecursive(String text, List<String> delimiters, int chunkSize, int level) {
```
Comment (yuye-aws) [inline src/main/java/org/opensearch/neuralsearch/processor/chunker/RecursiveCharacterChunker.java:72]: Either make this method static or not passing in class member variable **chunkSize** and **delimiters**

File: src/main/java/org/opensearch/neuralsearch/processor/chunker/RecursiveCharacterChunker.java [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+                    refined.add(chunk);
+                }
+            }
+            return refined;
+        } else {
+            return chunks.stream().filter(Objects::nonNull).toList();
+        }
+    }
+
+    private List<String> splitByLiteralDelimiter(String text, String delimiter) {
```
Comment (yuye-aws) [inline src/main/java/org/opensearch/neuralsearch/processor/chunker/RecursiveCharacterChunker.java:100]: Can be made static

File: src/main/java/org/opensearch/neuralsearch/processor/chunker/RecursiveCharacterChunker.java [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+                next = text.indexOf(delimiter, start);
+            }
+            if (start < text.length()) {
+                result.add(text.substring(start));
+            }
+        }
+        return result;
+    }
+
+    private List<String> mergeSplits(List<String> splits, String currentDelimiter, int chunkSize) {
```
Comment (yuye-aws) [inline src/main/java/org/opensearch/neuralsearch/processor/chunker/RecursiveCharacterChunker.java:125]: 1. Either static or no **chunkSize**
2. nit: currentDelimiter -> delimiter

**Comment by yuye-aws** [comment]
@YeonghyeonKO It's good to see you using an array of delimiters, but I'm wondering whether it works well with **max_chunk_limit** parameter. Suppose we get a string ["A,B.C,D"] and delimiters are [".", ","] and max_chunk_limit is 2. The expected output should be ["A,B", "C.D"].

In the chunk function, you are splitting across all delimiters and check max_chunk_limit later. The chunking speed would be sub-optimal as it is doing unnecessary chunking.

**Comment by yuye-aws** [comment]
Intuitively, you can do the chunk in an hierarchy manner. That is, you perform chunk on the 1st delimiter, then check the max_chunk_limit and if chunk_limit permits. You can then perform chunk on the 2nd delimiter.

Unfortunately, there would still be an issue on multiple documents inputs. To be specific, the input text field is a list of strings like ["A,B.C,D", "E,F.G,H"] and the delimiter is [".", ","] while max_chunk_limit is 5. The expected chunked segments should be ["A" ,"B", "C,D", "E.F", "G,H"], but from my understanding, the current code outputs  ["A" , "B", "C" ,"D", "E,F.G,H"]. The reason is that, when the chunker is performing chunk on the 1st string, it has no idea on the impact by the 2nd chunk.

**Comment by yuye-aws** [comment]
For this PR, a single workaround should be **only allow a single delimiter**. If you wish to perform chunking with multiple delimiters. You can **cascade** multiple processors like:

```
PUT _ingest/pipeline/text-chunking-ingest-pipeline
{
  "description": "A text chunking ingest pipeline using recursive character chunker",
  "processors": [
    {
      "text_chunking": {
        "algorithm": {
          "recursive_character": {
            "chunk_size": 50,
            "delimiters": ["\n\n",]
          }
        },
        "field_map": {
          "passage_text": "passage_chunk1"
        }
      }
    },
    {
      "text_chunking": {
        "algorithm": {
          "recursive_character": {
            "chunk_size": 50,
            "delimiters": [ "\n"]
          }
        },
        "field_map": {
          "passage_chunk1": "passage_chunk2"
        }
      }
    }
  ]
}
```

By doing so, it would be quite similar to the current delimiter algorithm. So I am going to suggest a simpler solution for you: introducing **chunk_size** parameter to the current delimiter chunker.

**Comment by yuye-aws** [comment]
@YeonghyeonKO There is no hurry on implement Integration tests. We can have a quick discussion on corner cases first.

**Comment by YeonghyeonKO** [comment]
@yuye-aws 
I agreed, there would be quite many corner cases while checking max_chunk_limit with multiple delimiters at once. Let's discuss whether it's possible to take a detour by introducing **chunk_size** parameter to the existing `delimiter` chunker.

**Comment by yuye-aws** [comment]
I tried to run your algorithm, but cannot start the neural search repo with simply **./gradlew run**. The error is similar to your CI results: https://github.com/opensearch-project/neural-search/actions/runs/14923033351/job/41922127280?pr=1312. I have created an issue to track this: https://github.com/opensearch-project/neural-search/issues/1313.

**Comment by YeonghyeonKO** [comment]
While it’s not an exact implementation of `RecursiveCharacterTextSplitter`, cascading multiple delimiter chunkers can serve as a lightweight alternative—so I think it makes sense to move forward with the #1330. Feel free to reopen this PR whoever wants to discuss #1261 again.

**Comment by YeonghyeonKO** [comment]
While it’s not an exact implementation of RecursiveCharacterTextSplitter, cascading multiple delimiter chunkers can serve as a lightweight alternative—so I think it makes sense to move forward with the https://github.com/opensearch-project/neural-search/pull/1330. Feel free to reopen this PR whoever wants to discuss https://github.com/opensearch-project/neural-search/issues/1261 again.
